hips/draw/healpix.py

```diff
@@ -88,3 +95,115 @@ def healpix_to_hips(hpx_data, tile_width, base_path, file_format='fits'):
 
     properties = HipsSurveyProperties(data=data)
     properties.write(base_path / 'properties')
+
+def hips_to_healpix_array(order: int):
+    """Create array giving mapping between:
+
+    HEALPix index ==> XY index
+    XY index ==> HEALPix index
+
+    Parameters
+    ----------
+    order : int
+        Order of HEALPix array.
+    """
+    def fill_array(npix, nsize, pos):
+        size = nsize ** 2
+        npix_idx = [[0 for i in range(size // 4)] for j in range(4)]
+        nb = [0 for i in range(4)]
+
+        for i in range(size):
+            if (i % nsize) < (nsize // 2):
+                dg = 0
+            else:
+                dg = 1
+
+            if i < (size / 2):
+                bh = 1
+            else:
+                bh = 0
+
+            quad = (dg << 1) | bh
+
+            if pos is None:
+                j = i
+            else:
+                j = pos[i]
+
+            npix[j] = npix[j] << 2 | quad
+            npix_idx[quad][nb[quad]] = j
+            nb[quad] += 1
+
+        if size > 4:
+            for i in range(4):
+                fill_array(npix, nsize // 2, npix_idx[i])
+
+    if order == 0:
+        xy_to_healpix = healpix_to_xy = [0]
+        return
+
+    nsize = 2 ** order
+    xy_to_healpix = [0 for i in range(nsize ** 2)]
+    healpix_to_xy = [0 for i in range(nsize ** 2)]
+
+    fill_array(xy_to_healpix, nsize, None)
+
+    for i in range(len(xy_to_healpix)):
+        healpix_to_xy[xy_to_healpix[i]] = i
+
+    return healpix_to_xy, xy_to_healpix
+
+def hips_to_healpix(hips_url: str, npix: int, hpx_output_path: str):
+    """Given a HiPS survey, generate a HEALPix map.
+
+    Parameters
+    ----------
+    hips_url : str
+        URL of HiPS survey.
+    npix : int
+        HEALPix pixel number.
+    hpx_output_path : str
+        HEALPix output path.
+    """
+    # Query the HiPS survey properties.
+    hips_properties = HipsSurveyProperties.fetch(hips_url + "/properties")
+
+    hips_tile_width = hips_properties.tile_width
+    order = int(np.log2(hips_tile_width))
+
+    healpix_to_xy, xy_to_healpix = hips_to_healpix_array(order)
+
+    healpix_map = np.empty(hp.nside2npix(2 ** 12), dtype=np.double) # 2 ** 12?
+    healpix_map[:] = hp.pixelfunc.UNSEEN
+
+    for ipix in range(npix):
```

Comment:
You should not have a double for loop.

The function should take a list (or iterable) of HiPSTile as input, and only have a loop over that.
The way you pre-allocate the output array before the loop loops OK.

There is no second loop over pixels within a tile, because the operations for those pixels are vectorised as Numpy expressions:
https://github.com/hipspy/hips/blob/0b52667f09fde5ea3d323b2ef7a6c2b0e064c2c8/hips/utils/healpix.py#L125